src/_pytest/doctest.py

```diff
@@ -368,10 +370,21 @@ def _patch_unwrap_mock_aware():
     else:
 
         def _mock_aware_unwrap(obj, stop=None):
-            if stop is None:
-                return real_unwrap(obj, stop=_is_mocked)
-            else:
-                return real_unwrap(obj, stop=lambda obj: _is_mocked(obj) or stop(obj))
+            try:
+                if stop is None or stop is _is_mocked:
+                    return real_unwrap(obj, stop=_is_mocked)
+                else:
+                    return real_unwrap(
+                        obj, stop=lambda obj: _is_mocked(obj) or stop(obj)
+                    )
+            except Exception as e:
+                warnings.warn(
+                    "Got %r when unwrapping %r.  This is usually caused "
+                    "by a violation of Python's object protocol; see e.g. "
+                    "https://github.com/pytest-dev/pytest/issues/5080" % (e, obj),
+                    PytestWarning,
+                )
+                raise
```

Comment:
How about making the code more readable in the first place?
```python
if stop is None:
    stop = _is_mocked
else:
    stop = lambda obj: _is_mocked(obj) or stop(obj)
return real_unwrap(obj, stop=stop)
```

```python
if stop is None:
    stop = _is_mocked
else:
    stop = lambda obj: _is_mocked(obj) or stop(obj)
return real_unwrap(obj, stop=stop)
```

Reply:
`flake8` error about assigning a lambda, and IMO defining a named function is less readable (and certainly longer).